synapse/handlers/sync.py

```diff
@@ -362,21 +362,36 @@ async def _wait_for_sync_for_user(
         # (since we now know that the device has received them)
         if since_token is not None:
             since_stream_id = since_token.to_device_key
-            # Delete device messages asynchronously and in batches using the task scheduler
-            await self._task_scheduler.schedule_task(
-                DELETE_DEVICE_MSGS_TASK_NAME,
-                resource_id=sync_config.device_id,
-                params={
-                    "user_id": sync_config.user.to_string(),
-                    "device_id": sync_config.device_id,
-                    "up_to_stream_id": since_stream_id,
-                },
+            # Fast path: delete a limited number of to-device messages up front.
+            # We do this to avoid the overhead of scheduling a task for every
```

Comment:
What does this overhead entail? I would have guessed we're adding a DB row and sending out a replication notification, which sounds lightweight---but haven't been following.

Other thoughts:
- should we have a metric for number of tasks spawned?
- for this to make sense we need to hit the fast past most of the time. How confident are we that this is the case?